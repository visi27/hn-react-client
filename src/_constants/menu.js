export const LINK_MENU_TYPE = 'LINK';
export const DROPDOWN_MENU_TYPE = 'DROPDOWN';
