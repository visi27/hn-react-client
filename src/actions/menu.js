import { MENU_LOAD } from '../constants/actionTypes';

export default function loadMenu() {
  return {
    type: MENU_LOAD,
  };
}
