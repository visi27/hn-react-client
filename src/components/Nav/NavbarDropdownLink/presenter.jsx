import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import v4 from 'uuid';

class NavbarDropdownLink extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false,
    };

    this.toggleDropdown = this.toggleDropdown.bind(this);
  }

  toggleDropdown(event) {
    event.preventDefault();
    this.setState({ show: !this.state.show });
  }

  render() {
    const show = this.state.show ? 'show' : '';
    const { dropdown } = this.props;
    const items = dropdown.children;
    return (
      <li className={`nav-item dropdown ${show}`}>
        <a
          className="nav-link dropdown-toggle"
          href={dropdown.href}
          id={dropdown.id}
          data-toggle="dropdown"
          aria-haspopup="true"
          aria-expanded="false"
          onClick={this.toggleDropdown}
        >
          {dropdown.title}
        </a>
        <div className={`dropdown-menu ${show}`} aria-labelledby={dropdown.id}>
          {items.map(item => (
            <a className="dropdown-item" href={item.href} key={v4()}>
              {item.title}
            </a>
          ))}
        </div>
      </li>
    );
  }
}

NavbarDropdownLink.propTypes = {
  dropdown: PropTypes.shape({
    title: PropTypes.string,
    href: PropTypes.string,
    type: PropTypes.string,
    id: PropTypes.string,
    children: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
      href: PropTypes.string,
    })),
  }),
};

NavbarDropdownLink.defaultProps = {
  dropdown: [],
};

export default NavbarDropdownLink;
