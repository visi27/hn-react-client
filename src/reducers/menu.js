import { MENU_LOAD } from '../constants/actionTypes';
import MenuConfig from '../config/menu';

const initialState = {
  elements: MenuConfig,
};

export default function (state = initialState, action) {
  switch (action.type) {
    case MENU_LOAD:
      return state;
    default:
      return state;
  }
}
