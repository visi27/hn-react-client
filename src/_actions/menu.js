import { MENU_LOAD } from '../_constants/actionTypes';

const loadMenu = () => ({
  type: MENU_LOAD,
});

export default loadMenu;
