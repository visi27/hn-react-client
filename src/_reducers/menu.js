import MenuConfig from '../_config/menu';
import { MENU_LOAD } from '../_constants/actionTypes';

const initialState = {
  elements: MenuConfig,
};

export default function (state = initialState, action) {
  switch (action.type) {
    case MENU_LOAD:
      return state;
    default:
      return state;
  }
}
