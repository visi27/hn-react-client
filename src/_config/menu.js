import * as menuTypes from '../_constants/menu';

export default [
  {
    title: 'Home',
    type: menuTypes.LINK_MENU_TYPE,
    href: '/',
  },
  {
    title: 'Latest',
    type: menuTypes.LINK_MENU_TYPE,
    href: '/latest',
    disabled: false,
  },
  {
    title: 'Ask HN',
    type: menuTypes.LINK_MENU_TYPE,
    href: '/ask',
    disabled: false,
  },
  {
    title: 'Show HN',
    type: menuTypes.LINK_MENU_TYPE,
    href: '/show',
    disabled: false,
  },
  {
    title: 'Favorites',
    type: menuTypes.LINK_MENU_TYPE,
    href: '/favorites',
    disabled: false,
  },
];
