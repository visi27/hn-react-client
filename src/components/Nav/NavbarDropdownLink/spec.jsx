import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDom from 'react-dom';
import renderer from 'react-test-renderer';
import * as menuTypes from '../../../_constants/menu';
import NavbarDropdownLink from './presenter';

configure({ adapter: new Adapter() });

describe('Nav', () => {
  const props = {
    dropdown: {
      title: 'Dropdown',
      type: menuTypes.DROPDOWN_MENU_TYPE,
      href: '#',
      id: 'dd01',
      children: [
        {
          title: 'SubMenu',
          type: menuTypes.LINK_MENU_TYPE,
          href: '#',
        },
        {
          title: 'Another SubMenu',
          type: menuTypes.LINK_MENU_TYPE,
          href: '#',
        },
      ],
    },
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDom.render(<NavbarDropdownLink {...props} />, div);
  });

  test('has valid snapshot', () => {
    const component = renderer.create(<NavbarDropdownLink {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows 2 dropdown items', () => {
    const element = mount(<NavbarDropdownLink {...props} />);
    expect(element.find('a.dropdown-item').length).toBe(2);
  });

  it('shows 1 dropdown menu ', () => {
    const element = mount(<NavbarDropdownLink {...props} />);
    expect(element.find('.nav-item .dropdown').length).toBe(1);
  });
});
