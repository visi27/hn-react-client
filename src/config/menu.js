import * as menuTypes from '../constants/menu';

export default [
  {
    title: 'Menu',
    type: menuTypes.LINK_MENU_TYPE,
    href: '#',
  },
  {
    title: 'Another Menu',
    type: menuTypes.LINK_MENU_TYPE,
    href: '#',
  },
  {
    title: 'Dropdown',
    type: menuTypes.DROPDOWN_MENU_TYPE,
    href: '#',
    id: 'dd01',
    children: [
      {
        title: 'SubMenu',
        type: menuTypes.LINK_MENU_TYPE,
        href: '#',
      },
      {
        title: 'Another SubMenu',
        type: menuTypes.LINK_MENU_TYPE,
        href: '#',
      },
    ],
  },
];
