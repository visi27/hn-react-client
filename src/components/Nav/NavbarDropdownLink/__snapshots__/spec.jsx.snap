// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav has valid snapshot 1`] = `
<li
  className="nav-item dropdown "
>
  <a
    aria-expanded="false"
    aria-haspopup="true"
    className="nav-link dropdown-toggle"
    data-toggle="dropdown"
    href="#"
    id="dd01"
    onClick={[Function]}
  >
    Dropdown
  </a>
  <div
    aria-labelledby="dd01"
    className="dropdown-menu "
  >
    <a
      className="dropdown-item"
      href="#"
    >
      SubMenu
    </a>
    <a
      className="dropdown-item"
      href="#"
    >
      Another SubMenu
    </a>
  </div>
</li>
`;
